File: csharp/HQNN.Core/QuantumNode.cs
using System;
using System.Numerics;
using System.Collections.Generic;
using System.Linq;

namespace HQNN.Core
{
    public class NodeConfig
    {
        public int NQubits { get; set; } = 2;
        public string InitialState { get; set; } = "superposition";
        public string DecoherenceModel { get; set; } = "lindblad";
        public double T1 { get; set; } = 100e-6;
        public double T2 { get; set; } = 50e-6;
    }

    public class QuantumNode
    {
        public int Id { get; private set; }
        public NodeConfig Config { get; private set; }
        public int NQubits { get; private set; }
        public int Dim { get; private set; }
        public Complex[,] DensityMatrix { get; private set; }
        public double Phase { get; set; }
        public double TopologicalHealth { get; set; }
        public int ErrorCount { get; set; }
        public List<double> MeasurementHistory { get; private set; }

        private static readonly Random Rng = new Random();

        public QuantumNode(int nodeId, NodeConfig config = null)
        {
            Id = nodeId;
            Config = config ?? new NodeConfig();
            NQubits = Config.NQubits;
            Dim = 1 << NQubits;
            DensityMatrix = InitializeDensityMatrix();
            Phase = Rng.NextDouble() * 2 * Math.PI;
            TopologicalHealth = 1.0;
            ErrorCount = 0;
            MeasurementHistory = new List<double>();
        }

        private Complex[,] InitializeDensityMatrix()
        {
            var state = new Complex[Dim];
            double norm = 1.0 / Math.Sqrt(Dim);

            switch (Config.InitialState)
            {
                case "superposition":
                    for (int i = 0; i < Dim; i++)
                        state[i] = new Complex(norm, 0);
                    break;
                case "zero":
                    state[0] = Complex.One;
                    break;
                case "ghz":
                    state[0] = new Complex(norm * Math.Sqrt(2), 0);
                    state[Dim - 1] = new Complex(norm * Math.Sqrt(2), 0);
                    break;
                default:
                    for (int i = 0; i < Dim; i++)
                        state[i] = new Complex(norm, 0);
                    break;
            }

            var dm = new Complex[Dim, Dim];
            for (int i = 0; i < Dim; i++)
                for (int j = 0; j < Dim; j++)
                    dm[i, j] = state[i] * Complex.Conjugate(state[j]);

            return dm;
        }

        public void ApplyGate(Complex[,] U)
        {
            if (U.GetLength(0) != Dim || U.GetLength(1) != Dim)
                throw new ArgumentException($"Gate dimension mismatch: expected {Dim}x{Dim}");

            var temp = MatMul(U, DensityMatrix);
            var Udag = ConjugateTranspose(U);
            DensityMatrix = MatMul(temp, Udag);
            Renormalize();
        }

        public void ApplyDecoherence(double gamma, double dt)
        {
            switch (Config.DecoherenceModel)
            {
                case "lindblad":
                    ApplyLindblad(gamma, dt);
                    break;
                case "phase_damping":
                    ApplyPhaseDamping(gamma, dt);
                    break;
                default:
                    ApplyLindblad(gamma, dt);
                    break;
            }
        }

        private void ApplyLindblad(double gamma, double dt)
        {
            var L = new Complex[Dim, Dim];
            for (int i = 0; i < Dim - 1; i++)
                L[i, i + 1] = new Complex(Math.Sqrt(gamma), 0);

            var Ldag = ConjugateTranspose(L);
            var LdagL = MatMul(Ldag, L);

            var LrhoLdag = MatMul(MatMul(L, DensityMatrix), Ldag);
            var LdagLrho = MatMul(LdagL, DensityMatrix);
            var rhoLdagL = MatMul(DensityMatrix, LdagL);

            for (int i = 0; i < Dim; i++)
                for (int j = 0; j < Dim; j++)
                {
                    var dissipator = LrhoLdag[i, j]
                        - 0.5 * LdagLrho[i, j]
                        - 0.5 * rhoLdagL[i, j];
                    DensityMatrix[i, j] += dissipator * dt;
                }

            Renormalize();
        }

        private void ApplyPhaseDamping(double gamma, double dt)
        {
            double p = 1.0 - Math.Exp(-gamma * dt);
            for (int i = 0; i < Dim; i++)
                for (int j = 0; j < Dim; j++)
                    if (i != j)
                        DensityMatrix[i, j] *= (1.0 - p);
            Renormalize();
        }

        private void Renormalize()
        {
            Complex trace = Complex.Zero;
            for (int i = 0; i < Dim; i++)
                trace += DensityMatrix[i, i];

            if (trace.Magnitude > 1e-12)
                for (int i = 0; i < Dim; i++)
                    for (int j = 0; j < Dim; j++)
                        DensityMatrix[i, j] /= trace;
        }

        public double GetPurity()
        {
            var rho2 = MatMul(DensityMatrix, DensityMatrix);
            double purity = 0.0;
            for (int i = 0; i < Dim; i++)
                purity += rho2[i, i].Real;
            return purity;
        }

        public double GetEntropy()
        {
            var eigenvalues = ComputeEigenvalues();
            double entropy = 0.0;
            foreach (var ev in eigenvalues)
            {
                double v = ev.Real;
                if (v > 1e-12)
                    entropy -= v * Math.Log2(v);
            }
            return entropy;
        }

        public double MeasureObservable(Complex[,] observable)
        {
            var product = MatMul(observable, DensityMatrix);
            double result = 0.0;
            for (int i = 0; i < Dim; i++)
                result += product[i, i].Real;
            MeasurementHistory.Add(result);
            return result;
        }

        private Complex[] ComputeEigenvalues()
        {
            var eigenvalues = new Complex[Dim];
            for (int i = 0; i < Dim; i++)
                eigenvalues[i] = new Complex(DensityMatrix[i, i].Real, 0);
            return eigenvalues;
        }

        public static Complex[,] MatMul(Complex[,] A, Complex[,] B)
        {
            int m = A.GetLength(0), n = B.GetLength(1), k = A.GetLength(1);
            var C = new Complex[m, n];
            for (int i = 0; i < m; i++)
                for (int j = 0; j < n; j++)
                    for (int l = 0; l < k; l++)
                        C[i, j] += A[i, l] * B[l, j];
            return C;
        }

        public static Complex[,] ConjugateTranspose(Complex[,] M)
        {
            int m = M.GetLength(0), n = M.GetLength(1);
            var R = new Complex[n, m];
            for (int i = 0; i < m; i++)
                for (int j = 0; j < n; j++)
                    R[j, i] = Complex.Conjugate(M[i, j]);
            return R;
        }

        public void Reset()
        {
            DensityMatrix = InitializeDensityMatrix();
            TopologicalHealth = 1.0;
            ErrorCount = 0;
            MeasurementHistory.Clear();
        }
    }
}

File: hqnn/core/quantum_node.py
import numpy as np
from scipy.linalg import expm
from dataclasses import dataclass, field
from typing import Optional, List
from hqnn.utils.gates import von_neumann_entropy, fidelity


@dataclass
class NodeConfig:
    n_qubits: int = 2
    initial_state: str = "superposition"
    decoherence_model: str = "lindblad"
    t1: float = 100e-6
    t2: float = 50e-6


class QuantumNode:
    def __init__(self, node_id: int, config: Optional[NodeConfig] = None):
        self.id = node_id
        self.config = config or NodeConfig()
        self.n_qubits = self.config.n_qubits
        self.dim = 2 ** self.n_qubits

        self.density_matrix = self._initialize_density_matrix()
        self.phase = np.random.uniform(0, 2 * np.pi)
        self.topological_health = 1.0
        self.coherence_time = 0.0
        self.error_count = 0
        self.measurement_history: List[float] = []

    def _initialize_density_matrix(self) -> np.ndarray:
        if self.config.initial_state == "superposition":
            state = np.ones(self.dim, dtype=complex) / np.sqrt(self.dim)
        elif self.config.initial_state == "zero":
            state = np.zeros(self.dim, dtype=complex)
            state[0] = 1.0
        elif self.config.initial_state == "ghz":
            state = np.zeros(self.dim, dtype=complex)
            state[0] = 1 / np.sqrt(2)
            state[-1] = 1 / np.sqrt(2)
        else:
            state = np.ones(self.dim, dtype=complex) / np.sqrt(self.dim)
        return np.outer(state, state.conj())

    def apply_gate(self, U: np.ndarray) -> None:
        assert U.shape == (self.dim, self.dim), f"Gate shape mismatch: {U.shape} vs {(self.dim, self.dim)}"
        self.density_matrix = U @ self.density_matrix @ U.conj().T
        self._renormalize()

    def apply_decoherence(self, gamma: float, dt: float) -> None:
        if self.config.decoherence_model == "lindblad":
            self._apply_lindblad(gamma, dt)
        elif self.config.decoherence_model == "amplitude_damping":
            self._apply_amplitude_damping(gamma, dt)
        elif self.config.decoherence_model == "phase_damping":
            self._apply_phase_damping(gamma, dt)

    def _apply_lindblad(self, gamma: float, dt: float) -> None:
        L = np.zeros((self.dim, self.dim), dtype=complex)
        for i in range(self.dim - 1):
            L[i, i + 1] = np.sqrt(gamma)
        Ldag = L.conj().T
        dissipator = (L @ self.density_matrix @ Ldag
                      - 0.5 * (Ldag @ L @ self.density_matrix
                                + self.density_matrix @ Ldag @ L))
        self.density_matrix += dissipator * dt
        self._renormalize()

    def _apply_amplitude_damping(self, gamma: float, dt: float) -> None:
        p = 1 - np.exp(-gamma * dt)
        K0 = np.array([[1, 0], [0, np.sqrt(1 - p)]], dtype=complex)
        K1 = np.array([[0, np.sqrt(p)], [0, 0]], dtype=complex)
        if self.dim == 2:
            rho_new = K0 @ self.density_matrix @ K0.conj().T + K1 @ self.density_matrix @ K1.conj().T
            self.density_matrix = rho_new
        else:
            self._apply_lindblad(gamma, dt)
        self._renormalize()

    def _apply_phase_damping(self, gamma: float, dt: float) -> None:
        p = 1 - np.exp(-gamma * dt)
        for i in range(self.dim):
            for j in range(self.dim):
                if i != j:
                    self.density_matrix[i, j] *= (1 - p)
        self._renormalize()

    def _renormalize(self) -> None:
        trace = np.trace(self.density_matrix)
        if np.abs(trace) > 1e-12:
            self.density_matrix /= trace
        self.density_matrix = (self.density_matrix + self.density_matrix.conj().T) / 2

    def measure_observable(self, observable: np.ndarray) -> float:
        expectation = float(np.real(np.trace(observable @ self.density_matrix)))
        self.measurement_history.append(expectation)
        return expectation

    def get_entropy(self) -> float:
        return von_neumann_entropy(self.density_matrix)

    def get_purity(self) -> float:
        return float(np.real(np.trace(self.density_matrix @ self.density_matrix)))

    def get_bloch_vector(self) -> np.ndarray:
        if self.dim != 2:
            return np.zeros(3)
        X = np.array([[0, 1], [1, 0]], dtype=complex)
        Y = np.array([[0, -1j], [1j, 0]], dtype=complex)
        Z = np.array([[1, 0], [0, -1]], dtype=complex)
        return np.array([
            float(np.real(np.trace(X @ self.density_matrix))),
            float(np.real(np.trace(Y @ self.density_matrix))),
            float(np.real(np.trace(Z @ self.density_matrix))),
        ])

    def reset(self) -> None:
        self.density_matrix = self._initialize_density_matrix()
        self.topological_health = 1.0
        self.error_count = 0
        self.measurement_history.clear()

    def clone(self) -> "QuantumNode":
        node = QuantumNode(self.id, self.config)
        node.density_matrix = self.density_matrix.copy()
        node.phase = self.phase
        node.topological_health = self.topological_health
        return node